client/src/entrypoints/admin/userbar.js

```diff
@@ -1,253 +1,348 @@
-  // source: https://stackoverflow.com/questions/1599660/which-html-elements-can-receive-focus
-  const focusableItemSelector = `a[href],
-    button:not([disabled]),
-    input:not([disabled])`;
-
-  // eslint-disable-next-line @typescript-eslint/no-use-before-define
-  trigger.addEventListener('click', toggleUserbar, false);
-
-  // make sure userbar is hidden when navigating back
-  // eslint-disable-next-line @typescript-eslint/no-use-before-define
-  window.addEventListener('pageshow', hideUserbar, false);
-
-  // Handle keyboard events on the trigger
-  // eslint-disable-next-line @typescript-eslint/no-use-before-define
-  userbar.addEventListener('keydown', handleTriggerKeyDown);
-  // eslint-disable-next-line @typescript-eslint/no-use-before-define
-  list.addEventListener('focusout', handleFocusChange);
-
-  // eslint-disable-next-line @typescript-eslint/no-use-before-define
-  resetItemsTabIndex(); // On initialisation, all menu items should be disabled for roving tab index
-
-  function showUserbar(shouldFocus) {
-    userbar.classList.add(isActiveClass);
-    trigger.setAttribute('aria-expanded', 'true');
-    // eslint-disable-next-line @typescript-eslint/no-use-before-define
-    list.addEventListener('click', sandboxClick, false);
+/**
+ * Tippy.js annotation as a custom element.
+ */
+class AxeAnnotation extends HTMLElement {
+  connectedCallback() {
+    const template = document.getElementById(
+      'wagtail-annotation-template',
+    ).content;
+    const shadowRoot = this.attachShadow({
+      mode: 'open',
+      // delegatesFocus: true,
+    });
+    shadowRoot.appendChild(template.cloneNode(true));
+
+    const btn = document.createElement('button');
+    btn.className = 'wagtail-annotation__toggle';
+    btn.innerText = 'Error';
+
+    btn.style.display = 'none';
+    shadowRoot.appendChild(btn);
+
+    const content = document.createElement('div');
+    content.innerHTML = this.dataset.message;
+
+    const tippyOptions = {
+      content: content,
+      trigger: 'click',
+      interactive: true,
+      theme: 'dropdown',
+      placement: 'bottom',
+      appendTo: () => shadowRoot,
+    };
+
+    console.log(tippy(btn, tippyOptions));
+  }
+}
+
+class Userbar extends HTMLElement {
+  constructor() {
+    super();
+  }
+
+  connectedCallback() {
+    const template = document.getElementById(
+      'wagtail-userbar-template',
+    ).content;
+    const shadowRoot = this.attachShadow({
+      mode: 'open',
+      // delegatesFocus: true,
+    });
+    shadowRoot.appendChild(template.cloneNode(true));
+
+    const userbar = shadowRoot.querySelector('[data-wagtail-userbar]');
+    const trigger = userbar.querySelector('[data-wagtail-userbar-trigger]');
+    this.trigger = trigger;
+    const list = userbar.querySelector('[role=menu]');
+    const listItems = list.querySelectorAll('li');
+    const isActiveClass = 'is-active';
+
+    userbar.style.display = 'none';
```

Comment:
Those are the most important changes to the userbar JS to make it work within a web component.